// jsdoc-3-docs-project-tools/SiteBuilder.js
/*jslint indent: 4, maxerr: 50, white: true, node: true, stupid: true */

/**
 * @file This file contains code to build html files from articles formatted with our custom mustache tags.
 * @author <a href="matthewkastor@gmail.com">Matthew Kastor</a>
 * @version 20121030
 * @requires fs
 * @requires events
 * @requires util
 * @requires path
 * @requires atropa-various-utilities
 * @requires ArticleParser
 * @requires mustacheFunctions
 * @requires apiGenerator
 * @requires utilities
 * @exports SiteBuilder
 */

'use strict';

var fs,
    events,
    util,
    path,
    directory,
    utilities,
    articleParser,
    mustacheFunctions,
    generator;

fs     = require('fs');
events = require('events');
util   = require('util');
path   = require('path');

/** @see atropa-various-utilities */
directory        = require('atropa-various-utilities').directory;
/** @see utilities */
utilities = require('./utilities');
/** @see ArticleParser */
articleParser    = new (require('./ArticleParser').ArticleParser)();
/** @see mustacheFunctions */
mustacheFunctions= require('./mustacheFunctions');
/** @see apiGenerator */
generator        = new (require('./ApiGenerator').ApiGenerator)();

/**
 * Represents the jsdoc 3 website builder.
 * @class
 */
function SiteBuilder() {
    

    
    /** Holds the array of articles */
    this.articles = [];
    
    /**
     * The source directory where article files are found.
     * @default <code> Jake/articles </code>
     */
    this.srcdir = 'Jake/articles/';
    
    /**
     * The directory to put the generated site pages into.
     * @default <code> ./ </code>
     */
    this.outdir = './';
}

/**
 * Purges directories before regenerating updated files.
 */
SiteBuilder.prototype.preClean = function preClean() {
    console.log('Cleaning pages...');
    [
        this.outdir,
        'Jake/API/describeTags',
        'Jake/API/jsdoc/jsdoc.js',
        'Jake/API/jsdoc/node_modules',
        'Jake/API/jsdoc/plugins',
        'Jake/API/jsdoc/rhino_modules',
        'Jake/API/jsdoc/templates'
    ].forEach(function(where) {
        directory.purgeHtmlPages(where);
    });
};

/**
 * Generates missing tag articles and extended documents files for new tags.
 */
SiteBuilder.prototype.generateMissingTagsArticles = function generateMissingTagsArticles() {
    utilities.generateMissingTagsArticles();
};

SiteBuilder.prototype.parseArticles = function parseArticles() {
    console.log('Building index...');
    articleParser.loadArticleFiles(this.srcdir);
    this.articles = articleParser.getArticles();
};

/**
 * Builds all site pages from sources.
 */
SiteBuilder.prototype.buildSitePages = function buildSitePages() {
    var mustachio;
    this.parseArticles();
    console.log('Building new pages...');
    mustachio = new mustacheFunctions.MustacheComb();
    mustachio.generateArticles(this.articles, this.outdir);
};

/**
 * Runs all functions to completely regenerate the site.
 */
SiteBuilder.prototype.buildSite = function buildSite() {
    var my = this;
    console.log('Building site...');
    this.preClean();
    generator.generateDocs(function(){
        my.generateMissingTagsArticles();
        my.buildSitePages();
    });
};

exports.SiteBuilder = SiteBuilder;
